clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii.hs

```diff
@@ -0,0 +1,151 @@
+{-|
+  Copyright   :  (C) 2024, Google LLC
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+ Contains a wrapper for the Xilinx GMII to SGMII PMA core :
+  "LogiCORE IP Ethernet 1000BASE-X PCS/PMA or SGMII"
+  https://www.xilinx.com/xsw/gig_ethernet_pcs_pma
+
+-}
+
+{-# LANGAUGE CPP #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=10 #-}
+
+module Clash.Cores.Xilinx.Ethernet.Gmii where
+
+import Clash.Explicit.Prelude
+
+import Clash.Annotations.Primitive
+import Clash.Cores.Xilinx.Ethernet.Gmii.BlackBoxes
+import Clash.Cores.Xilinx.Ethernet.Gmii.Types
+import Clash.Signal.Internal(ClockN, DiffClock(..))
+
+-- | Wrapper for the LogiCORE IP Ethernet 1000BASE-X PCS/PMA or SGMII.
+-- Configured to function as a GMII to SGMII bridge.
+gmiiSgmiiBridge ::
+  forall async sgmii625 gmii125 .
+  ( KnownDomain sgmii625
+  , KnownDomain gmii125
+  , DomainPeriod sgmii625 ~  Picoseconds 1600
+  , DomainPeriod gmii125 ~ Nanoseconds 8
+  , DomainActiveEdge sgmii625 ~ 'Rising
+  , DomainActiveEdge gmii125 ~ 'Rising
+  , DomainResetKind async ~ 'Asynchronous
+  , DomainResetKind gmii125 ~ 'Synchronous
+  , DomainResetPolarity async ~ 'ActiveHigh
+  , DomainResetPolarity gmii125 ~ 'ActiveHigh
+  ) =>
+  -- | Reference clock coming from the PHY.
+  DiffClock sgmii625 ->
+  -- | Asynchronous reset for the bridge.
+  Reset async ->
+  -- | Signal detect from the PHY. Either connect to the PHY's signal detect or use
+  -- a constant @True@, otherwise the link will never come up.
+  Signal async Bool ->
```

Comment:
Ah wait, no, that doesn't work for `Reset gmii125` because it already has a synchronous reset (output).

The primitive is rather like our clock generator functions: they get a clock input and a reset that is so asynchronous it isn't even synchronously deasserted, and they output a clock with a reset (that reset, in the case of the clock generators, is optionally synchronous).

The type signature for the clock generators makes sense here as well:

```haskell
DiffClock sgmii625 -> Reset sgmii625 -> ... -> ...{Clock gmii125, Reset gmii125, ...}
```

So both `refRst` and `signalDetect` should be `sgmii625`, IMO.

Reply:
The reason why I made the `async` domain is because `refRst` and `signalDetect` are not associated with any domain according to the documentation (as indicated by `n/a` under `Clock Domain` in table 2 of DS264.

It's truly a shame we don't have a straightforward way to create truly asynchronous signals at the moment. 
I do feel like associating them with `sgmii625` gives off the false sense that they are synchronized to that domain.

Though if you feel strongly on this, I'm willing to annotate them that way. I also do see that a polymorphic `async` domain that falsely relates a reset and a signal also may cause problems.

Reply:
Asynchronous signals on the output of a function are the truly insidious case; those functions probably all need to have the word _unsafe_ in their name. It is way too easy to mishandle those signals. Luckily, here, `signalDetect` is an input; the primitive doesn't care any way whatsoever. I suspect the `signalDetect` signal will be synchronised inside the primitive, and it wouldn't surprise me if it even didn't care about any glitches. Ethernet links don't go up and down that quickly. And it will probably just come from an FPGA pin connected to the PHY, without any user-defined circuitry in Clash.

So I think it's not super critical what domain you declare on `signalDetect`. If it were up to me, I'd declare it `sgmii625` with some documentation that the signal need not be synchronised. But you could keep the `async` domain.

For `refRst`, I really feel `Reset sgmii625` makes a lot of sense. We already declare it asynchronous (and active high). Clash doesn't differentiate between asynchronous and synchronous deassertion; in Clash, the usual behaviour of an asynchronous reset is that it asserts asynchronously but deasserts synchronously. Like for the clock generators, that's not the case here, and like the clock generators, the primitive's timing is determined by a clock input with a reset input that is completely asynchronous. Since we have the `HasAsynchronousReset`, it makes a lot of sense to put the domain `sgmii625` on that reset input. The output clock and reset are derived from the input clock and reset. They are pretty much grouped.

Reply:
I'll change the domains to `sgmii625`